src/Npgsql/NpgsqlParameter.cs

```diff
@@ -368,7 +368,22 @@ public NpgsqlDbType NpgsqlDbType
             {
                 if (_dataTypeName != null)
                     return _dataTypeName;
-                throw new NotImplementedException("Infer from others");
+                if (_npgsqlDbType.HasValue)
+                {
+                    var primitiveNpgsqlType = _npgsqlDbType.Value & ~NpgsqlDbType.Array & ~NpgsqlDbType.Range;
+                    var primitiveTypeName = GlobalTypeMapper.Instance.ToPgTypeName(primitiveNpgsqlType);
+
+                    if ((_npgsqlDbType.Value & NpgsqlDbType.Array) != 0 ||
+                        (_npgsqlDbType.Value & NpgsqlDbType.Range) != 0)
+                    {
+                        primitiveTypeName += "[]";
```

Comment:
This should be done in ToPgTypeName and not here, just like NpgsqlDbType.Array is added when inferring the NpgsqlDbType.

Reply:
Code moved to ToPgTypeName in 3b88b4a. Thanks for the advice!